File: billbull-frontend/src/utils/accountCodeGenerator.js
/**
 * Client-side mirror of the backend AccountCodeGenerator
 * (financials/chartofaccounts/AccountCodeGenerator.java).
 *
 * Used only for the "Auto NNNN" preview in the account modals — the authoritative
 * allocation happens server-side (blank code on POST /api/ledger/accounts, or
 * GET /api/ledger/accounts/next-code). Keep the two in sync when bands change.
 *
 * Rules:
 *  1. With a parent, children are numbered inside the parent's hundred-block
 *     (6000 -> 6001..6099, 1050 -> 1051..1099).
 *  2. Otherwise (or if that block is full) fall back to the bands the group really
 *     owns — Expenses also owns 6000-6999 and 7500-7999, Income also owns 7000-7499.
 *  3. Always take the FIRST FREE code in a band, never max+1 (max+1 walks straight
 *     out of the band once a code like the seeded 5999 exists, and collides with 6000).
 */

const GROUP_BANDS = {
  assets: [[1000, 1999]],
  asset: [[1000, 1999]],
  liabilities: [[2000, 2999]],
  liability: [[2000, 2999]],
  equity: [[3000, 3999]],
  income: [[4000, 4999], [7000, 7499]],
  revenue: [[4000, 4999], [7000, 7499]],
  expenses: [[5000, 5999], [6000, 6999], [7500, 7999]],
  expense: [[5000, 5999], [6000, 6999], [7500, 7999]]
};

const FALLBACK_BANDS = [[9000, 9999]];

const parseCode = (code) => {
  if (code === null || code === undefined) return null;
  const trimmed = String(code).trim();
  if (!trimmed) return null;
  const parsed = Number.parseInt(trimmed, 10);
  return Number.isFinite(parsed) && String(parsed) === trimmed ? parsed : null;
};

const bandsForGroup = (group) =>
  GROUP_BANDS[String(group || '').trim().toLowerCase()] || FALLBACK_BANDS;

const childBandFor = (parentCode) => {
  const parent = parseCode(parentCode);
  if (parent === null) return null;
  const blockStart = Math.floor(parent / 100) * 100;
  return [Math.max(blockStart, parent) + 1, blockStart + 99];
};

const firstFree = ([min, max], used) => {
  for (let candidate = min; candidate <= max; candidate += 1) {
    if (!used.has(candidate)) return candidate;
  }
  return null;
};

/**
 * @returns {string} the next free code, or '' when every candidate band is full
 *                   (the user must then type a code manually).
 */
export const generateNextAccountCode = ({ group, parentCode, existingAccounts = [] }) => {
  const used = new Set(
    existingAccounts
      .map(account => parseCode(account?.code))
      .filter(code => code !== null)
  );

  const parentBand = childBandFor(parentCode);
  if (parentBand) {
    const allocated = firstFree(parentBand, used);
    if (allocated !== null) return String(allocated);
  }

  for (const [min, max] of bandsForGroup(group)) {
    // Skip the band root itself (5000, 6000, …) — those are the group/header accounts.
    const allocated = firstFree([min + 1, max], used);
    if (allocated !== null) return String(allocated);
  }

  return '';
};

export default generateNextAccountCode;

File: billbull-backend/src/main/java/com/billbull/backend/financials/chartofaccounts/AccountCodeGenerator.java
package com.billbull.backend.financials.chartofaccounts;

import java.util.Collection;
import java.util.List;
import java.util.Locale;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * Single source of truth for allocating the next free Chart of Accounts code.
 *
 * <p>Two rules, in order:
 * <ol>
 *   <li><b>Parent-derived</b> — when a parent account is given, children are numbered inside the
 *       parent's own hundred-block (parent {@code 6000} → {@code 6001..6099}, parent {@code 1050} →
 *       {@code 1051..1099}). This keeps the code aligned with the tree position, which the seeded
 *       COA already assumes (6001-6099 under 6000, 4101+ under 4100).</li>
 *   <li><b>Group bands</b> — no parent (or the parent block is exhausted) falls back to the bands
 *       the seeded COA actually occupies. Note these are <i>not</i> one 1000-wide band per group:
 *       Expenses also owns 6000-6999 and 7500-7999, Income also owns 7000-7499.</li>
 * </ol>
 *
 * <p>Allocation always scans for the <b>first free</b> code in a band rather than {@code max + 1},
 * so a code sitting at the top of a band (e.g. the seeded {@code 5999 Rounding Adjustment}) can
 * never push the next allocation out of the band and onto an existing account.
 */
public final class AccountCodeGenerator {

    /** A contiguous, inclusive range of numeric account codes. */
    public record CodeBand(int min, int max) {}

    /** Bands owned by each account group, in allocation preference order. */
    private static List<CodeBand> bandsForGroup(String accountGroup) {
        String group = accountGroup == null ? "" : accountGroup.trim().toLowerCase(Locale.ROOT);
        return switch (group) {
            case "assets", "asset"           -> List.of(new CodeBand(1000, 1999));
            case "liabilities", "liability"  -> List.of(new CodeBand(2000, 2999));
            case "equity"                    -> List.of(new CodeBand(3000, 3999));
            case "income", "revenue"         -> List.of(new CodeBand(4000, 4999), new CodeBand(7000, 7499));
            case "expenses", "expense"       -> List.of(new CodeBand(5000, 5999),
                                                        new CodeBand(6000, 6999),
                                                        new CodeBand(7500, 7999));
            // Unknown/custom groups get a dedicated band so they can never collide with the standard COA.
            default                          -> List.of(new CodeBand(9000, 9999));
        };
    }

    /**
     * The hundred-block a child of {@code parentCode} belongs in — e.g. 6000 → 6001..6099,
     * 1050 → 1051..1099. Returns {@code null} when the parent code is not a plain number.
     */
    static CodeBand childBandFor(String parentCode) {
        Integer parent = parseCode(parentCode);
        if (parent == null) return null;
        int blockStart = (parent / 100) * 100;
        // Start above the parent itself so a parent sitting mid-block (1050) never re-issues its own code.
        return new CodeBand(Math.max(blockStart, parent) + 1, blockStart + 99);
    }

    /**
     * Allocates the first free code for a new account.
     *
     * @param parentCode      selected parent account code, may be {@code null}/blank
     * @param accountGroup    root group name ("Assets", "Expenses", …)
     * @param existingCodes   every code already present in the COA (any format; non-numeric ignored)
     * @return the allocated code, or {@code null} if every candidate band is full
     */
    public static String nextCode(String parentCode, String accountGroup, Collection<String> existingCodes) {
        Set<Integer> used = existingCodes.stream()
                .map(AccountCodeGenerator::parseCode)
                .filter(java.util.Objects::nonNull)
                .collect(Collectors.toSet());

        CodeBand parentBand = childBandFor(parentCode);
        if (parentBand != null) {
            Integer allocated = firstFree(parentBand, used);
            if (allocated != null) return String.valueOf(allocated);
        }

        for (CodeBand band : bandsForGroup(accountGroup)) {
            // Skip the band root itself (5000, 6000, …) — those are the group/header accounts.
            Integer allocated = firstFree(new CodeBand(band.min() + 1, band.max()), used);
            if (allocated != null) return String.valueOf(allocated);
        }

        return null;
    }

    private static Integer firstFree(CodeBand band, Set<Integer> used) {
        for (int candidate = band.min(); candidate <= band.max(); candidate++) {
            if (!used.contains(candidate)) return candidate;
        }
        return null;
    }

    /** Parses a plain numeric code; returns {@code null} for blank or non-numeric codes. */
    private static Integer parseCode(String code) {
        if (code == null) return null;
        String trimmed = code.trim();
        if (trimmed.isEmpty()) return null;
        try {
            return Integer.valueOf(trimmed);
        } catch (NumberFormatException ex) {
            return null;
        }
    }

    private AccountCodeGenerator() {}
}
